### PR TITLE
Updates of ESmry

### PR DESCRIPTION
**Pull request by tskille**
- new overload function for keywordList, list keys based on pattern (string argument)
 - summary program wild card elements (keys including * or ?)
 - python bindings to new member function

### CODE REVIEW

**Comment by bska** [comment]
jenkins build this please

File: python/cxx/eclipse_io.cpp [hunk 1]
```diff
@@ -341,7 +341,10 @@ void python::common::export_IO(py::module& m) {
         .def("__get_all", &get_smry_vector)
         .def("__get_at_rstep", &get_smry_vector_at_rsteps)
         .def_property_readonly("start_date", &esmry_start_date)
-        .def_property_readonly("list_of_keys", &Opm::EclIO::ESmry::keywordList);
+        .def_property_readonly("list_of_keys", (const std::vector<std::string>& (Opm::EclIO::ESmry::*) (void) const)
+            &Opm::EclIO::ESmry::keywordList)
+        .def("list_of_keys_pattern", (std::vector<std::string> (Opm::EclIO::ESmry::*) (const std::string&) const)
+            &Opm::EclIO::ESmry::keywordList);
```
Comment (joakim-hove) [inline python/cxx/eclipse_io.cpp:347]: In addition to this C++ binding code I would recommend a small pure python function in the `__init__.py` 
```Python
def keys(self, pattern=None):
      if pattern is None:
            return self.list_of_keys
      else:
            return self.list_of_keys_pattern( pattern )
```

**Comment by joakim-hove** [comment]
Valuable addition!

File: python/cxx/eclipse_io.cpp (hunk 1 above)
Comment (joakim-hove) [reply python/cxx/eclipse_io.cpp:347]: If you are willing to call the `list_of_keys` as a *function* - you will get this for free by just using the same name in the binding code.

Comment (tskille) [reply python/cxx/eclipse_io.cpp:345]: I changed the bindings code to two functions of keys. 

```C++
        .def("keys", (const std::vector<std::string>& (Opm::EclIO::ESmry::*) (void) const)
            &Opm::EclIO::ESmry::keywordList)
        .def("keys", (std::vector<std::string> (Opm::EclIO::ESmry::*) (const std::string&) const)
            &Opm::EclIO::ESmry::keywordList);
```

This is in line with what you are suggesting, right ? 

Having **list_of_keys** as a property seemed like a good idea when I didn't have the overloaded function with the pattern argument. 

Thanks @joakim-hove, I like this better.

Comment (joakim-hove) [reply python/cxx/eclipse_io.cpp:345]: > This is in line with what you are suggesting, right ?

yes!

**Comment by bska** [comment]
jenkins build this please

**Pull request merged**